src/sys/sendfile.rs

```diff
@@ -120,6 +120,38 @@ cfg_if! {
                 )
             }
         }
+    } else if #[cfg(solarish)] {
+        #[derive(Debug, Copy, Clone)]
+        /// Mapping of the raw C sendfilevec_t struct
+        pub struct SendfileVec(libc::sendfilevec_t);
+
+        const SFV_FD_SELF: i32 = -2;
+
+        impl SendfileVec {
+            /// initialises SendfileVec to send data directly from the process's address space
+            /// same in C with sfv_fd set to SFV_FD_SELF.
+            pub fn newself(
+                off: off_t,
+                len: usize
+            ) -> SendfileVec {
+                SendfileVec(libc::sendfilevec_t{sfv_fd: SFV_FD_SELF, sfv_flag: 0, sfv_off: off, sfv_len: len})
+            }
+
+            /// initialises SendfileVec to send data from `fd`.
+            pub fn new<F: AsFd>(
+                fd: F,
```

Comment:
This still doesn't capture anything about fd's lifetime.  With this API, it's possible to use a RawFd after the file has been closed.  That's why I suggested using BorrowedFd.  You might be able to do it, without additional runtime cost, like this:
```rust
pub struct SendfileVec<'a> {
    raw: libc::sendfilevec_t,
    phantom: PhantomData<BorrowedFd<'a>>
}
pub fn new<F: AsFd, 'a>(fd: &'a F, off: off_t, len: usize) -> Self<'a> {
    let bfd = fd.as_fd();
    Self{raw: libc::sendfilevec_t{sfv_fd: bfd.as_raw_fd(), ...}, phantom: PhantomData}}
}
```